kroxylicious-kms-provider-aws-kms/src/main/java/io/kroxylicious/kms/provider/aws/kms/AwsV4SigningHttpRequestBuilder.java

```diff
@@ -0,0 +1,364 @@
+    private StringToSignResult computeStringToSign(CanonicalRequestResult canonicalRequestResult, String isoDateTime, String isoDate) {
+        var stringToSignLines = new ArrayList<String>();
+        stringToSignLines.add("AWS4-HMAC-SHA256");
+        stringToSignLines.add(isoDateTime);
+        var credentialScope = isoDate + "/" + region + "/" + service + "/" + AWS_4_REQUEST;
+        stringToSignLines.add(credentialScope);
+        stringToSignLines.add(canonicalRequestResult.canonicalRequestHash());
+        var stringToSign = String.join("\n", stringToSignLines);
+        return new StringToSignResult(credentialScope, stringToSign);
+    }
+
+    private record StringToSignResult(String credentialScope, String stringToSign) {}
+
+    @NonNull
+    private String computeAuthorization(CanonicalRequestResult canonicalRequestResult, StringToSignResult stringToSignResult, String isoDate) {
+        var dateHmac = hmac(("AWS4" + secretKey).getBytes(StandardCharsets.UTF_8), isoDate);
+        var regionHmac = hmac(dateHmac, this.region);
+        var serviceHmac = hmac(regionHmac, this.service);
+        var signHmac = hmac(serviceHmac, AWS_4_REQUEST);
+        var signature = HEX_FORMATTER.formatHex(hmac(signHmac, stringToSignResult.stringToSign()));
+
+        return "AWS4-HMAC-SHA256 Credential=" + accessKey + "/" + stringToSignResult.credentialScope() + ", SignedHeaders=" + canonicalRequestResult.signedHeaders()
+                + ", Signature=" + signature;
+    }
+
+    /**
+     * This implementation must match the HTTP client's computation of the contents of the Host header.
+     *
+     * @param uri uri
+     * @return host string
+     */
+    @VisibleForTesting
+    String getHostHeaderForSigning(URI uri) {
+        int port = uri.getPort();
+        String host = uri.getHost();
+
+        boolean defaultPort;
+        if (port == -1) {
+            defaultPort = true;
+        }
+        else if (uri.getScheme().toLowerCase(Locale.ROOT).equals("https")) {
+            defaultPort = port == 443;
+        }
+        else {
+            defaultPort = port == 80;
+        }
+
+        if (defaultPort) {
+            return host;
+        }
+        else {
+            return host + ":" + port;
+        }
+    }
+
+    private static String normalizeSpaces(String value) {
+        return value.replaceAll("\\s+", " ").trim();
+    }
+
+    private static byte[] sha256(byte[] bytes) {
+        var digest = getSha256Digester();
+        digest.update(bytes);
+        return digest.digest();
+    }
+
+    private static byte[] hmac(byte[] key, String msg) {
+        try {
+            var mac = getHmacSHA256();
+            mac.init(new SecretKeySpec(key, "HmacSHA256"));
+            return mac.doFinal(msg.getBytes(StandardCharsets.UTF_8));
+        }
+        catch (InvalidKeyException e) {
+            throw new KmsException("Failed to initialize hmac", e);
+        }
+    }
+
+    @NonNull
+    private static Mac getHmacSHA256() {
```

Comment:
I'm sure JCE have their reasons for with mixed CamelCASE, but I think this would be better named `newHmacSha256()` (since elsewhere you're using AwsKms etc, and generally we're strictly camelcase in Kroxy).